### PR TITLE
Expose notification event for receive pipeline completed

### PR DESCRIPTION
**Pull request by sfarmar**
Needed to move performance counters to a NServiceBus.WindowsPerformanceCounters package #747

### CODE REVIEW

**Comment by andreasohlund** [comment]
Will making the actual events we need to expose public be a separate PR? (like https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/Transports/ReceivePipelineCompleted.cs )

**Comment by andreasohlund** [comment]
Marking as WIP for now, @sfarmar will add more to this PR

**Comment by andreasohlund** [comment]
@Particular/nservicebus-maintainers after talking to @danielmarbach we decided to go with a specific api for the receive pipeline event since we believe this is the way to go with future events.

@SimonCropp you advocated this when I proposed a generic `.Notification()` api a while back

You where right, I was wrong :)

**Comment by andreasohlund** [comment]
Note that we might have to expose the `NotificationSubscriptions` in the future if downstream components would like to add their own events but that is YAGNI for now

File: src/NServiceBus.AcceptanceTests/Notifications/When_subscribed_to_ReceivePipelineCompleted.cs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    using AcceptanceTesting;
+    using Configuration.AdvanceExtensibility;
+    using EndpointTemplates;
+    using Features;
+    using NUnit.Framework;
+
+    class When_subscribed_to_ReceivePipelineCompleted
+    {
+        [Test]
+        public async Task Should_receive_then_notification()
```
Comment (timbussmann) [inline src/NServiceBus.AcceptanceTests/Notifications/When_subscribed_to_ReceivePipelineCompleted.cs]: the_notification?

File: src/NServiceBus.AcceptanceTests/Notifications/When_subscribed_to_ReceivePipelineCompleted.cs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+        class TestEndpoint : EndpointConfigurationBuilder
+        {
+            public TestEndpoint()
+            {
+                EndpointSetup<DefaultServer>(c =>
+                {
+                    c.Pipeline.OnReceivePipelineCompleted(e =>
+                    {
+                        var testContext = (Context) c.GetSettings().Get<ScenarioContext>();
```
Comment (timbussmann) [inline src/NServiceBus.AcceptanceTests/Notifications/When_subscribed_to_ReceivePipelineCompleted.cs]: why not use the context parameter in `EndpointSetup`? (`EndpointSetup<DefaultServer>((config, context) =>`)

File: src/NServiceBus.Core/Pipeline/Incoming/ReceivePipelineCompleted.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+        /// <summary>
+        /// Time when the receive pipline completed.
+        /// </summary>
+        public DateTime CompletedAt { get; }
+
+        /// <summary>
+        /// Constructs the event.
+        /// </summary>
+        public ReceivePipelineCompleted(IncomingMessage processedMessage, DateTime startedAt, DateTime completedAt)
```
Comment (timbussmann) [inline src/NServiceBus.Core/Pipeline/Incoming/ReceivePipelineCompleted.cs]: ctor should be at the top?

File: src/NServiceBus.Core/Pipeline/Incoming/ReceivePipelineConfigExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+    /// <summary>
+    /// Provides access to receive pipeline notifications.
+    /// </summary>
+    public static class ReceivePipelineConfigExtensions
+    {
+        /// <summary>
+        /// Subscribes to notifications for completed receive pipelines.
+        /// </summary>
+        public static void OnReceivePipelineCompleted(this PipelineSettings pipelineSettings, Func<ReceivePipelineCompleted, Task> subscription)
```
Comment (timbussmann) [inline src/NServiceBus.Core/Pipeline/Incoming/ReceivePipelineConfigExtensions.cs]: not against that approach, but I dislike that we now have two different public event subscription APIs (well, three including the legacy API). So there are some events registered via `config.Pipeline` and some via `config.Notifications` ? 👎

Comment (andreasohlund) [inline src/NServiceBus.Core/Pipeline/Incoming/ReceivePipelineConfigExtensions.cs]: Synced with @timbussmann and decided to go ahead with this but to align the error notifications in a separate PR

Comment (danielmarbach) [inline src/NServiceBus.Core/Pipeline/Incoming/ReceivePipelineConfigExtensions.cs]: @timbussmann shouldn't walk his 🐶 when we have important discussions ;)

File: src/NServiceBus.AcceptanceTests/Notifications/When_subscribed_to_ReceivePipelineCompleted.cs (hunk 1 above)
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/Notifications/When_subscribed_to_ReceivePipelineCompleted.cs]: How about `Should_receive_notifications`

File: src/NServiceBus.AcceptanceTests/Notifications/When_subscribed_to_ReceivePipelineCompleted.cs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        }
+
+        class Context : ScenarioContext
+        {
+            public bool ReceivePipelineCompletedFired { get; set; }
+            public ReceivePipelineCompleted ReceivePipelineCompletedMessage { get; set; }
+            public string MessageId { get; set; }
+        }
+
+        class TestEndpoint : EndpointConfigurationBuilder
```
Comment (danielmarbach) [inline src/NServiceBus.AcceptanceTests/Notifications/When_subscribed_to_ReceivePipelineCompleted.cs]: How about `SubscribingEndpoint` or `SubscribedEndpoint`

File: src/NServiceBus.Core.Tests/API/APIApprovals.ApproveNServiceBus.approved.txt [hunk 6]
```diff
@@ -2480,7 +2491,7 @@ namespace NServiceBus.Pipeline
     {
         public PipelineExecutor() { }
     }
-    public class PipelineSettings
+    public class PipelineSettings : NServiceBus.Configuration.AdvanceExtensibility.ExposeSettings
```
Comment (danielmarbach) [inline src/NServiceBus.Core.Tests/API/APIApprovals.ApproveNServiceBus.approved.txt]: According to https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md this should be fine

> Introducing a new base class
> So long as it does not introduce any new abstract members or change the semantics or behavior of existing members, a type can be introduced into a hierarchy between two existing types. For example, between .NET Framework 1.1 and .NET Framework 2.0, we introduced DbConnection as a new base class for SqlConnection which previously derived from Component.

File: src/NServiceBus.Core/Pipeline/Incoming/ReceivePipelineCompleted.cs [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    /// </summary>
+    public class ReceivePipelineCompleted
+    {
+        /// <summary>
+        /// The message that was processed.
+        /// </summary>
+        public IncomingMessage ProcessedMessage { get; }
+
+        /// <summary>
+        /// Time when the receive pipline started.
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Pipeline/Incoming/ReceivePipelineCompleted.cs]: Spelling in pipeline

File: src/NServiceBus.Core/Pipeline/Incoming/ReceivePipelineCompleted.cs [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        /// </summary>
+        public DateTime StartedAt { get; }
+
+        /// <summary>
+        /// Time when the receive pipline completed.
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Pipeline/Incoming/ReceivePipelineCompleted.cs]: Spelling in pipeline

File: src/NServiceBus.Core/Pipeline/Incoming/ReceivePipelineCompleted.cs [hunk 9]
```diff
@@ -0,0 +1,40 @@
+namespace NServiceBus
+{
+    using System;
+    using Transport;
+
+    /// <summary>
+    /// The ReceivePipeline completed event.
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Pipeline/Incoming/ReceivePipelineCompleted.cs]: How about something

> Indicates when a receive pipeline is completed. 

Should we also indicate in the doco here that this event is only raised for the shared queue and not or the instance specific one?

File: src/NServiceBus.AcceptanceTests/Notifications/When_subscribed_to_ReceivePipelineCompleted.cs (hunk 2 above)
Comment (andreasohlund) [inline src/NServiceBus.AcceptanceTests/Notifications/When_subscribed_to_ReceivePipelineCompleted.cs]: Not following? (the signature is `Action<EndpointConfiguration, RunDescriptor>` )

File: src/NServiceBus.Core/Pipeline/Incoming/ReceivePipelineCompleted.cs (hunk 9 above)
Comment (andreasohlund) [inline src/NServiceBus.Core/Pipeline/Incoming/ReceivePipelineCompleted.cs]: I don't think this is correct, the instance specific queue is using the mainPipelineExecutor as well so the event would fire there as well?

https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/StartableEndpoint.cs#L145

**Comment by andreasohlund** [comment]
Review comments addressed

**Comment by danielmarbach** [comment]
@andreasohlund ok, wouldn't we want then the pipeline id in the event as well?

**Comment by andreasohlund** [comment]
>ok, wouldn't we want then the pipeline id in the event as well?

Sounds resonable but lets handle that in a separate pull?

**Comment by danielmarbach** [comment]
Ok for me

**Comment by andreasohlund** [comment]
We can discuss the use of contexts outside of this PR. Merging this

**Pull request merged**